gossipdns/dns.go

```diff
@@ -0,0 +1,239 @@
+package gossipdns
+
+import (
+	"bytes"
+	"fmt"
+	"math/rand"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/miekg/dns"
+
+	. "github.com/weaveworks/weave/common"
+	"github.com/weaveworks/weave/net/address"
+)
+
+const (
+	topDomain        = "."
+	reverseDNSdomain = "in-addr.arpa."
+	etcResolvConf    = "/etc/resolv.conf"
+	udpBuffSize      = uint16(4096)
+)
+
+type DNSServer struct {
+	ns     *Nameserver
+	domain string
+	ttl    uint32
+	port   int
+
+	servers   []*dns.Server
+	upstream  *dns.ClientConfig
+	tcpClient *dns.Client
+	udpClient *dns.Client
+}
+
+func NewDNSServer(ns *Nameserver, domain string, port int, ttl uint32,
+	clientTimeout time.Duration) (*DNSServer, error) {
+
+	s := &DNSServer{
+		ns:        ns,
+		domain:    domain,
+		ttl:       ttl,
+		port:      port,
+		tcpClient: &dns.Client{Net: "tcp", ReadTimeout: clientTimeout},
+		udpClient: &dns.Client{Net: "udp", ReadTimeout: clientTimeout, UDPSize: udpBuffSize},
+	}
+	var err error
+	if s.upstream, err = dns.ClientConfigFromFile(etcResolvConf); err != nil {
```

Comment:
Integration tests are not a substitute of good unit tests. Otherwise, we could just `rm -rf */*_test.go`... ;)

We currently test things like _"a truncated UDP response falls back to a TCP query"_. I would like to see things like that in the DNS unit tests...